src/rgw/rgw_op.cc

```diff
@@ -1793,6 +1793,146 @@ void RGWStatBucket::execute()
   }
 }
 
+int RGWGetBucketOplog::verify_permission()
+{
+  if (s->user->user_id.compare(s->bucket_owner.get_id()) != 0)
+    return -EACCES;
+
+  return 0;
+}
+
+void RGWGetBucketOplog::execute()
+{
+  RGWAccessHandle h;
+  struct rgw_log_entry entry;
+  string st = s->info.args.get("start-time"),
+         et = s->info.args.get("end-time"),
+         marker = s->info.args.get("marker");
+  uint64_t epoch = 0;
+  utime_t  ut_st, ut_et, ut_curr, ut_end, ut_marker;
+  uint32_t skip = 0, count = 0, index;
+  string oid_marker;
+  struct tm bdt;
+  time_t tt;
+  bool found = false;
+
+  http_ret = 0;
+  if (st.empty() || et.empty()) {
+    ldout(s->cct, 0) << "start-time and end-time must be non-null! " << st << dendl;
+    http_ret = -EINVAL;
+    return;
+  }
+
+  if (utime_t::parse_date(st, &epoch, NULL) < 0) {
+    ldout(s->cct, 0) << "Error parsing date. st=" << st << dendl;
+    http_ret = -EINVAL;
+    return;
+  }
+  ut_st = utime_t(epoch, 0);
+
+  if (utime_t::parse_date(et, &epoch, NULL) < 0) {
+    ldout(s->cct, 0) << "Error parsing date. et=" << et << dendl;
+    http_ret = -EINVAL;
+    return;
+  }
+  ut_et = utime_t(epoch, 0);
+
+  if (ut_st >= ut_et || (ut_et - ut_st) > utime_t(24 * 3600, 0)) {
+    ldout(s->cct, 0) << "Invalid argument! st=" << st << " et=" << et << dendl;
+    http_ret = -EINVAL;
+    return;
+  }
+
+  if (!marker.empty()) {
+    size_t pos = marker.find('#');
+    if (pos == std::string::npos) {
+      ldout(s->cct, 0) << "marker is wrong! marker:" << marker << dendl;
+      http_ret = -EINVAL;
+      return;
+    }
+    oid_marker = marker.substr(0, pos);
+
+    pos += 1;
+    string skip_str = marker.substr(pos, marker.length() - pos);
+    int ret = stringtoul(skip_str, &skip);
+    if (ret < 0) {
```

Comment:
Because some entries might have the same timestamp in a log object, I cannot confirm which entry as start entry only according to timestamp in the next HTTP request.
The marker compose of log object id and sequences number now, and the sequences number is as the last entry in the last log object in the previous HTTP response, and I can know the start entry according to log object id and sequence number in the marker.